server/server.go

```diff
@@ -134,74 +150,148 @@ func (s *RPCServer) ListenAndServe() {
+	if err != nil {
+		return err
+	}
+	defer outputCh.Close()
+
+	// Setup a WaitGroup for use by consume(). This WaitGroup will be 0
+	// when all consumers are finished consuming messages.
+	consumersWg := sync.WaitGroup{}
+
+	// consumerTags is used when we later want to tell AMQP that we want to cancel our consumers.
+	consumerTags := []string{}
 	for _, binding := range s.bindings {
-		err := s.consume(binding, inputCh)
+		consumerTag, err := s.consume(binding, inputCh, consumersWg)
 		if err != nil {
 			return err
 		}
-	}
 
-	outputCh, err := conn.Channel()
-	if err != nil {
-		return err
+		consumerTags = append(consumerTags, consumerTag)
 	}
 
-	defer outputCh.Close()
+	// This WaitGroup will reach 0 when the responder() has finished sending all responses.
+	responderWg := sync.WaitGroup{}
+
+	go s.responder(outputCh, responderWg)
+
+	inputConnClosed := inputConn.NotifyClose(make(chan *amqp.Error))
+	outputConnClosed := outputConn.NotifyClose(make(chan *amqp.Error))
+
+	// Wait for graceful shutdown.
+	select {
+	case <-s.stopChan:
+		// 1. Tell amqp we want to shut down by cancelling all the consumers.
+		for _, consumerTag := range consumerTags {
+			err = inputCh.Cancel(consumerTag, false)
+			if err != nil {
+				return err
+			}
+		}
+
+		// 2. We've told amqp to stop delivering messages, now we wait for all
+		// the consumers to finish inflight messages.
+		consumersWg.Wait()
 
-	go s.responder(outputCh)
+		// 3. Close the responses chan and wait until the consumers are finished.
+		// We might still have responses we want to send.
+		close(s.responses)
+		responderWg.Wait()
 
-	err, ok := <-conn.NotifyClose(make(chan *amqp.Error))
+		// 4. We have no more messages incoming and we've sent all our responses.
+		inputConn.Close()
+		outputConn.Close()
+
+	// Check to see if the connections are closed without us calling .Close() first.
+	// This counts as an error even if AMQP thinks this was a graceful shutdown.
+	case err, ok := <-inputConnClosed:
+		if !ok {
+			return ErrUnexpectedConnClosed
+		}
+		return err
+	case err, ok := <-outputConnClosed:
+		if !ok {
+			// If we get here the connection was closed without calling Close().
+			// This counts as an error even if AMQP thinks this was a graceful shutdown.
+			return ErrUnexpectedConnClosed
+		}
+		return err
+	}
+
+	// Wait for the disconnection to complete after we have told the server to gracefully close.
+	err, ok := <-inputConnClosed
 	if !ok {
 		// The connection was closed gracefully.
 		return nil
 	}
-	// The connection wasn't closed gracefully.
+	// Something happend while closing the connection.
 	return err
 }
 
-func (s *RPCServer) consume(binding HandlerBinding, inputCh *amqp.Channel) error {
+func (s *RPCServer) consume(binding HandlerBinding, inputCh *amqp.Channel, wg sync.WaitGroup) (string, error) {
 	queueName, err := s.declareAndBind(inputCh, binding)
 	if err != nil {
-		return err
+		return "", err
 	}
 
+	consumerTag := uuid.Must(uuid.NewV4()).String()
```

Comment:
No way to reach the auto generated consumer tag when the default value (`""`) is used? Is UUID really the preferred way to do this?

Reply:
In that case we can't know the consumer tag until the first message has arrived.